[pt1][quant] Add fbgemm_qlinear_dynamic op

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23330 [pt1][quant] Fix the dimension mismatch issues when running the BERT model
* #23212 [pt1][quant] Add the type matching rule for qconfig_dict
* #23128 [pt1][quant] Dynamic Quantized Linear operator and module
* **#23104 [pt1][quant] Add fbgemm_qlinear_dynamic op**

As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
[WIP][RFC][pt1][quant] Add fbgemm_qlinear_dynamic op

As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[WIP][RFC][pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[WIP][RFC][pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[WIP][RFC][pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[WIP][RFC][pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -273,6 +273,49 @@ def test_cat(self, X, num, axis, relu):
+        if use_relu:
+            Y_fp32_ref[Y_fp32_ref < 0.0] = 0.0
+
+        torch.testing.assert_allclose(Y_fp32, Y_fp32_ref, rtol=0.0001, atol=1e-3)
```
Can we have  a test that is exact: If you quantize and dequantize the activations prior to F.linear, can you get an exact match?

jianyuh replied:
Right: currently we use a special example (the same as https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L11177-L11198), so we can have the exact match.

raghuramank100 reviewed (changes requested):
Please address comments, thanks

jianyuh committed:
Update on "[pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh commented:
> Please address comments, thanks

Updated @raghuramank100 .

raghuramank100 reviewed (approved):
Please create a follow up task to add randomized testing for this op.

jianyuh committed:
Update on "[pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh commented:
Created the task T47798162 for the unit test for the exact match between static and dynamic quantization.

jianyuh commented:
> Please create a follow up task to add randomized testing for this op.

create a unit test for the exact match between static and dynamic quantization.

jianyuh committed:
Update on "[pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)

jianyuh committed:
Update on "[pt1][quant] Add fbgemm_qlinear_dynamic op"


As suggested in https://github.com/pytorch/pytorch/pull/22891, we will add an overload for ```torch.fbgemm_linear_int8_weight``` (dynamic quantized version of linear function) that takes PackedLinearWeight as input and is pretty much the same in signature as regular aten::linear.

Differential Revision: [D16381552](https://our.internmc.facebook.com/intern/diff/D16381552/)
